src/main.cpp

```diff
@@ -0,0 +1,416 @@
+                  std::istream_iterator<std::string>(),
+                  std::back_inserter(args_));
+    }
+
+    bool empty() const override {
+        return empty_;
+    }
+
+    void stop() override {
+        empty_ = true;
+    }
+
+    CmdRaw get_next_cmd() override {
+        empty_ = true;
+        return {cmd_, args_};
+    }
+
+private:
+    bool empty_{false};
+
+    std::string cmd_;
+    std::vector<std::string> args_;
+
+};
+
+
+
+// stub test class
+class CmdParserStub : public ICmdParser {
+private:
+    constexpr static size_t kMaxQueryCount{3};
+    size_t query_count{0}; 
+
+public:
+    explicit CmdParserStub([[maybe_unused]] std::string& cmd_line) {}
+
+    bool empty() const override {
+        return empty_;
+    }
+
+
+    void stop() override { empty_ = true; }
+
+
+    CmdRaw get_next_cmd() override {
+        if (query_count >= kMaxQueryCount) {
+            return {{"exit"}};
+        }
+
+        ++query_count;
+        return {{"test"}, {"arg1", "arg2_" + std::to_string(query_count)}};
+    }
+
+private:
+    bool empty_{false};
+};
+
+
+
+// ----------------
+// Exceptions
+// ----------------
+
+struct SubcommandParseException : public std::exception {
+    char const* what() const noexcept override {
+        return "Failed to parse command!";
+    }
+};
+
+struct CmdNotFoundException : public std::exception {
+    char const* what() const noexcept override {
+        return "Command is not found!";
+    }
+};
+
+struct CmdErrorException : public std::exception {
+    char const* what() const noexcept override {
+        return "Error was raised during command execution!";
+    }
+};
+
+struct CmdExitException : public std::exception {
+    char const* what() const noexcept override {
+        return "Execution finished using exit command!";
+    }
+};
+
+
+
+// ----------------
+// Commands to execute
+// ----------------
+
+// REDO: singleton
+class IExecutable {
+public:
+    virtual void execute(Args& args, IStreams& streams, State& state) = 0;
+    virtual ~IExecutable() = default;
+};
+
+// test
+class TestCmd : public IExecutable {
+public:
+    void execute(Args& args, IStreams& streams, [[maybe_unused]] State& state) {
+        auto& cout_ = streams.get_out_stream();
+        cout_ << "TestCmd running with arguments: ";
+        for (const auto& arg : args.get_all()) {
+           cout_ << arg << " "; // fix
+        }
+        cout_ << std::endl;
+    }
+};
+
+
+// echo
+class EchoCmd : public IExecutable {
+public:
+    void execute(Args& args, IStreams& streams, [[maybe_unused]] State& state) {
+        streams.get_out_stream() << args.get_arg(0) << std::endl;
+    }
+};
+
+
+// exit
+class ExitCmd : public IExecutable {
+public:
+    void execute([[maybe_unused]] Args& args, [[maybe_unused]] IStreams& streams, [[maybe_unused]] State& state) {
+        throw CmdExitException();
+    }
+};
+
+
+
+
+// ----------------
+// Program class
+// ----------------
+
+// pimpl
+class Program {
```

Comment:
Should we split Program onto two parts: command-name vs pointer-to-function map and .run()?